### ISSUE TITLE
The affinity of a worker process is reset after torch.randperm is called

### ISSUE THREAD

**Issue by ludovicdelfau** [pytorch/pytorch#28389]
## 🐛 Bug

The affinity of a worker process is reset after `torch.randperm` is called in the worker process.

## To Reproduce

1. Change the worker process affinity in the `worker_init_fn` function.
1. Call `torch.randperm` in the `__iter__` method of an IterableDataser.
1. Now the worker process affinity is back to the original one.

The following code sample reproduce the problem :

```
import os
import torch
import torch.utils.data

class TestDataset(torch.utils.data.IterableDataset):
    
    def __iter__(self):
        assert(len(os.sched_getaffinity(0)) == 1)
        torch.randperm(1)
        assert(len(os.sched_getaffinity(0)) == 1)
        return range(1)

def worker_init_fn(_):
    os.sched_setaffinity(0, list(os.sched_getaffinity(0))[0:1])
        
if __name__ == '__main__':
    assert(len(os.sched_getaffinity(0)) > 1)
    dataset = TestDataset()
    dataloader = torch.utils.data.DataLoader(dataset, num_workers=1, worker_init_fn=worker_init_fn)
    for _ in dataloader:
        pass
```

## Expected behavior

I except that a call to `torch.randperm` inside a worker process doesn't touch the process affinity.

## Environment

I use the Docker image `pytorch/pytorch:1.3-cuda10.1-cudnn7-runtime`.

```
Collecting environment information...
PyTorch version: 1.3.0
Is debug build: No
CUDA used to build PyTorch: 10.1.243

OS: Ubuntu 16.04.6 LTS
GCC version: (Ubuntu 5.4.0-6ubuntu1~16.04.11) 5.4.0 20160609
CMake version: Could not collect

Python version: 3.6
Is CUDA available: No
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA

Versions of relevant libraries:
[pip] numpy==1.17.2
[pip] torch==1.3.0
[pip] torchvision==0.4.1a0+d94043a
[conda] blas                      1.0                         mkl  
[conda] mkl                       2019.4                      243  
[conda] mkl-service               2.3.0            py36he904b0f_0  
[conda] mkl_fft                   1.0.14           py36ha843d7b_0  
[conda] mkl_random                1.1.0            py36hd6b4f25_0  
[conda] pytorch                   1.3.0           py3.6_cuda10.1.243_cudnn7.6.3_0    pytorch
[conda] torchvision               0.4.1                py36_cu101    pytorch
```

cc @ezyang @gchanan @zou3519 @jerryzh168 @SsnL

**Comment by ezyang** [comment]
cc @ilia-cher

**Comment by gchanan** [comment]
Do we know if randperm is important here?  What if we use some other torch function?

**Comment by LudovicDelfau** [comment]
> 
> 
> Do we know if randperm is important here? What if we use some other torch function?

With these functions, I got the same error :

* torch.zeros(1)
* torch.get_num_threads()

But with these functions, I didn't get the error :

* torch.empty(1)
* torch.can_cast(torch.double, torch.float)
* torch.seed()

**Comment by peterbell10** [comment]
It looks like this is the default behaviour of intel's OpenMP implementation. I say this for three reasons:

1. I can only reproduce the issue on platforms with MKL/intel OpenMP
2. Setting the environment variable `KMP_AFFINITY=disabled` fixes the issue
3. Setting `KMP_AFFINITY=verbose` explicity shows openmp setting the affinity

```
OMP: Info #250: KMP_AFFINITY: pid 648 tid 648 thread 0 bound to OS proc set 0-7
OMP: Info #250: KMP_AFFINITY: pid 651 tid 651 thread 0 bound to OS proc set 0-7
```

In normal python code I see `kmp` doing this setup when `torch` is imported . However, in in the data loader (or any code using `multiprocessing`) it is delayed until the first function call that uses OpenMP. Adding a dummy call to `torch.get_num_threads` before the call to `os.sched_setaffinity` also works around the issue.

So, should we add a dummy OpenMP call into `DataLoader` as a workaround?

**Comment by ilia-cher** [comment]
I don't think we provide guarantees on thread affinity when using intra- or inter-op parallelism as of the moment. Technically, thread may change its core/processor depending on external factors (parallelization library used; afaik in some cases OS may choose to move threads, etc). Some affinity (e.g. usage of specific CPU) may be enforced on the process level (e.g. numactl).

**Comment by SsnL** [comment]
I wonder if things are different with fork vs spawn.

**Comment by peterbell10** [comment]
The following code works correctly with `spawn` or `forkserver` but not with `fork`:

```python
import os
import torch
import multiprocessing as mp

def test(a):
    os.sched_setaffinity(0, [2])
    torch.randperm(1)
    return os.sched_getaffinity(0)

if __name__ == '__main__':
    mp.set_start_method('spawn') # or 'forkserver'
    with mp.Pool(2) as p:
        print(p.map(test, range(2)))
```

I'm assuming that `kmp` lazily recreates its thread pool on the first OpenMP call and that's also when the affinity is set.

**Comment by SsnL** [comment]
Then we can easily mp.util.register_after_fork to fix this.
